staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go

```diff
@@ -531,6 +773,390 @@ func TestNoOpUpdates(t *testing.T) {
 	}
 }
 
+func TestStoreUpdateHooks(t *testing.T) {
+	// To track which hooks were called in what order.  Not all hooks can
+	// mutate the object.
+	var milestones []string
+
+	setAnn := func(obj runtime.Object, key string) {
+		pod := obj.(*example.Pod)
+		if pod.Annotations == nil {
+			pod.Annotations = make(map[string]string)
+		}
+		pod.Annotations[key] = "true"
+	}
+	mile := func(s string) {
+		milestones = append(milestones, s)
+	}
+
+	testCases := []struct {
```

Comment:
I thought one of the semantics you wanted was that every attempt at an update got a separate finish call.  I don't think I see a test case checking for that semantic.  Given the likelihood of a refactor on the update code, I think it could help ensure the behavior sticks.  Is there a reasonable way to have the `NewTestGenericStoreRegistry` fail the guaranteed update?

Reply:
Look at TestStoreUpdateHooksInnerRetry() - it uses the TTLFunc hook to force a failure and retry on the inner loop, so you see cases like:

```
expectMilestones: []string{"BeginUpdate", "TTLError", "FinishUpdate(false)", "BeginUpdate", "TTL", "FinishUpdate(true)", "AfterUpdate", "Decorator"},
```

and 

```
expectMilestones: []string{"BeginUpdate", "TTLError", "FinishUpdate(false)", "BeginUpdate", "TTLError", "FinishUpdate(false)"},
```

I assume that handles what you are asking for?  Will resolve on that presumption.

Reply:
> I assume that handles what you are asking for? Will resolve on that presumption.

so it is.  thank you.